### PR TITLE
Add functional tests and integration tests for Waiters

### PR DESCRIPTION
**Pull request by Quanzzzz**
Add integration tests for DynamoDB Waiters;
Add functional tests for RestJsonWithWaiters Waiters;

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/waiters/WaitersSyncFunctionalTest.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+import software.amazon.awssdk.core.retry.backoff.BackoffStrategy;
+import software.amazon.awssdk.core.waiters.PollingStrategy;
+import software.amazon.awssdk.core.waiters.WaiterResponse;
+import software.amazon.awssdk.http.SdkHttpResponse;
+import software.amazon.awssdk.services.restjsonwithwaiters.RestJsonWithWaitersClient;
+import software.amazon.awssdk.services.restjsonwithwaiters.model.AllTypesRequest;
+import software.amazon.awssdk.services.restjsonwithwaiters.model.AllTypesResponse;
+import software.amazon.awssdk.services.restjsonwithwaiters.waiters.RestJsonWithWaitersWaiter;
+
+public class WaitersSyncFunctionalTest {
```
Comment (dagnir) [inline test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/waiters/WaitersSyncFunctionalTest.java:21]: Seems like we need to test other scenarios as well such as 

 - What happens when we reach max number of attempts
 - What happens when we get an unexpected exception
 - What happens to the client and scheduled executor when the the waiter is closed?
 - etc.

cc: @zoewangg for more suggestions

Comment (Quanzzzz) [reply test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/waiters/WaitersSyncFunctionalTest.java:40]: The max number of attempts case is added in sync Waiters. Unexpected exception case has been covered by the tests for generic Waiters. For the third case, if the client and scheduled executor are configured while creating the Waiters then they won't be closed while the Waiters are closed, and tests related to this case are also added.

**Review by dagnir — APPROVED** [review]

File: test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/waiters/WaitersSyncFunctionalTest.java (hunk 1 above)
Comment (zoewangg) [reply test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/waiters/WaitersSyncFunctionalTest.java:40]: @dagnir Good call! I added the test cases with unexpected exception and caught a bug in waiterRuntime 😅

Comment (Quanzzzz) [reply test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/waiters/WaitersSyncFunctionalTest.java:40]: 😅Guess that's my fault. I misunderstood the unmatched exception case in the `BaseWaiterTest`, thanks for making sure about it again.

**Pull request merged**